app/scripts/lib/transaction/smart-transactions.ts

```diff
@@ -0,0 +1,319 @@
+import SmartTransactionsController from '@metamask/smart-transactions-controller';
+import {
+  Fee,
+  Fees,
+  SmartTransactionStatuses,
+  SmartTransaction,
+} from '@metamask/smart-transactions-controller/dist/types';
+import type { Hex } from '@metamask/utils';
+import {
+  TransactionController,
+  TransactionMeta,
+  TransactionParams,
+} from '@metamask/transaction-controller';
+import log from 'loglevel';
+import {
+  RestrictedControllerMessenger,
+  EventConstraint,
+} from '@metamask/base-controller';
+import {
+  AddApprovalRequest,
+  UpdateRequestState,
+  StartFlow,
+  EndFlow,
+} from '@metamask/approval-controller';
+
+import { decimalToHex } from '../../../../shared/modules/conversion.utils';
+import { CANCEL_GAS_LIMIT_DEC } from '../../../../shared/constants/smartTransactions';
+import {
+  SMART_TRANSACTION_CONFIRMATION_TYPES,
+  ORIGIN_METAMASK,
+} from '../../../../shared/constants/app';
+
+const namespace = 'SmartTransactions';
+
+type AllowedActions =
+  | AddApprovalRequest
+  | UpdateRequestState
+  | StartFlow
+  | EndFlow;
+
+export type SmartTransactionsControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof namespace,
+    AllowedActions,
+    EventConstraint,
+    AllowedActions['type'],
+    never
+  >;
+
+export type FeatureFlags = {
+  extensionActive: boolean;
+  mobileActive: boolean;
+  smartTransactions: {
+    expectedDeadline?: number;
+    maxDeadline?: number;
+    returnTxHashAsap?: boolean;
+  };
+};
+
+export type SubmitSmartTransactionRequest = {
+  transactionMeta: TransactionMeta;
+  smartTransactionsController: SmartTransactionsController;
+  transactionController: TransactionController;
+  isSmartTransaction: boolean;
+  controllerMessenger: SmartTransactionsControllerMessenger;
+  featureFlags: FeatureFlags;
+};
+
+export class SmartTransactionHook {
+  #approvalFlowEnded: boolean;
+
+  #approvalFlowId: string;
+
+  #transactionMeta: TransactionMeta;
+
+  #smartTransactionsController: SmartTransactionsController;
+
+  #transactionController: TransactionController;
+
+  #isSmartTransaction: boolean;
+
+  #controllerMessenger: SmartTransactionsControllerMessenger;
+
+  #featureFlags: {
+    extensionActive: boolean;
+    mobileActive: boolean;
+    smartTransactions: {
+      expectedDeadline?: number;
+      maxDeadline?: number;
+      returnTxHashAsap?: boolean;
+    };
+  };
+
+  #isDapp: boolean;
+
+  #chainId: Hex;
+
+  #txParams: TransactionParams;
```

Comment:
Minor, we usually sort the properties in alphabetical order if there are many, just for easier readability and maintenance.